### PR TITLE
quilt features + encoding plots + weight init

### PR DESCRIPTION
**Pull request by ritvikvasan**
1. quilt dataloader
2. encoding greedy plots
3. rearranging and linting
4. some scripts

### CODE REVIEW

File: CVAE_testbed/datasets/quilt_aics_features.py [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+        self.BATCH_SIZE = BATCH_SIZE
+        self.corr = corr
+        self.shuffle = shuffle
+        self.model_kwargs = model_kwargs
+        self.train = train
+
+        Batches_X_train, Batches_C_train, Batches_conds_train = torch.empty([0]), torch.empty([0]), torch.empty([0])
+        Batches_X_test, Batches_C_test, Batches_conds_test = torch.empty([0]), torch.empty([0]), torch.empty([0])
+
+        ds = Package.browse("aics/pipeline_integrated_single_cell", "s3://allencell")
```
Comment (JacksonMaxfield) [inline CVAE_testbed/datasets/quilt_aics_features.py]: You should consider adding a `try` `except` for if the features have been downloaded or not. If the `json_quilt_path` doesn't exist, download them to that dir

Comment (JacksonMaxfield) [reply CVAE_testbed/datasets/quilt_aics_features.py]: This would be in place of the default `try` `except` below.

Comment (ritvikvasan) [reply CVAE_testbed/datasets/quilt_aics_features.py]: Good point! Will make change

File: CVAE_testbed/datasets/quilt_aics_features.py [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+        json_files = [pos_json for pos_json in os.listdir(path_to_json) if pos_json.endswith('.json')]
+
+        meta_to_file_name = []
+        for f in ds["cell_features"]:
+            meta_to_file_name.append({"filename": f, **ds["cell_features"][f].meta})
+        metas = pd.DataFrame(meta_to_file_name)
+
+        # Specify path to config file for FeatureDatabase
+        try:
+            db = FeatureDatabase(model_kwargs['config_path'])
```
Comment (JacksonMaxfield) [inline CVAE_testbed/datasets/quilt_aics_features.py]: Note to me: This makes me think I should really combine a lot of our various feature / data sets together into one big one, or at least, combine all the various integrated cell feature / data sets...

File: CVAE_testbed/datasets/quilt_aics_features.py [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+
+        semi = metas.merge(t, left_on="CellId", right_on="CellId", suffixes=("_meta", "_mito"))
+
+        # Only interphase or no interphase
+        semi['Interphase and Mitotic Stages [stage]'] = semi['Interphase and Mitotic Stages [stage]'].apply(lambda x: 0 if x == 0.0 else 1)
+
+        dd = defaultdict(list)
+        for i in range(len(semi['filename'])):
+            this_file = semi['filename'][i]
+            a = json.loads(open('/home/ritvik.vasan/test/' + this_file).read())
```
Comment (JacksonMaxfield) [inline CVAE_testbed/datasets/quilt_aics_features.py]: This will fail if someone uses the `model_kwargs['json_quilt_path']`, you create a variable up top when you check if the kwargs include a `json_quilt_path`, if they do, store it in a centralized variable, if it doesn't, default to something other than your specific user directory. I usually do `~/Desktop/{dataset_name}`.

File: CVAE_testbed/datasets/quilt_aics_features.py [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+            a = dict([(key, value) for key,value in a.items() if key not in ['imsize_orig', 'com', 'angle', 'flipdim', 'imsize_registered' ] ])
```
Comment (JacksonMaxfield) [inline CVAE_testbed/datasets/quilt_aics_features.py]: Can you explain what is going on here? I think there is a much simpler way to do this (if I am understanding correctly) but I don't really understand what is going on...

File: CVAE_testbed/datasets/quilt_aics_features.py [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+        categorical_features = ['Gene', 'ProteinDisplayName', 'StructureDisplayName']
+
+        categorical_dataframe = meta_plus_features[categorical_features]
+
+        non_categorical_dataframe = meta_plus_features[[c for c in meta_plus_features.columns if c not in categorical_features]]
+
+        one_hot_categorical_features = pd.get_dummies(categorical_dataframe, prefix = None, drop_first = True)
+        
+        num_of_cells = len(non_categorical_dataframe)
+        print(num_of_cells)
```
Comment (JacksonMaxfield) [inline CVAE_testbed/datasets/quilt_aics_features.py]: Should generally use `logging` over `print`

File: CVAE_testbed/datasets/quilt_aics_features.py [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+
+        # Here we normalize X with min max
+        # x = non_categorical_dataframe.values
+        # min_max_scaler = preprocessing.MinMaxScaler()
+        # x_scaled = min_max_scaler.fit_transform(x)
+        # non_categorical_dataframe = pd.DataFrame(x_scaled)
```
Comment (JacksonMaxfield) [inline CVAE_testbed/datasets/quilt_aics_features.py]: Why include this?

File: CVAE_testbed/datasets/quilt_aics_features.py [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+        # x_train_scaled = sklearn.preprocessing.normalize(x[:28000, 1:6], axis=0)
+        # x_test_scaled = sklearn.preprocessing.normalize(x[28000:, 1:6], axis=0)
+        # non_categorical_train = pd.DataFrame(x_train_scaled)
+        # non_categorical_test = pd.DataFrame(x_test_scaled)
+        # print(non_categorical_train.shape, non_categorical_test.shape )
+        # non_categorical_train = pd.DataFrame(np.concatenate((x[:28000, 0:1], x_train_scaled), axis = 1))
+        # non_categorical_test = pd.DataFrame(np.concatenate((x[28000:, 0:1], x_test_scaled), axis = 1))
+        
+        # non_categorical_dataframe = non_categorical_dataframe.drop_duplicates()
+        # non_categorical_dataframe = non_categorical_dataframe[(np.abs(stats.zscore(non_categorical_dataframe))<3).all(axis=1)]
```
Comment (JacksonMaxfield) [inline CVAE_testbed/datasets/quilt_aics_features.py]: Why include this?

File: CVAE_testbed/datasets/quilt_aics_features.py [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+        for j, i in enumerate(range(self.num_batches)):
+            X_train = all_non_categorical_X_train[i*self.BATCH_SIZE: (i+1)*self.BATCH_SIZE, :]
+            X_test = all_non_categorical_X_test[i*self.BATCH_SIZE: (i+1)*self.BATCH_SIZE, :]
+            X_categorical = all_categorical_X[i*self.BATCH_SIZE: (i+1)*self.BATCH_SIZE, :]
+
+            # X = torch.cat((X, X_categorical), 1)
+
+            if X_train.size()[0] != self.BATCH_SIZE:
+                break
+            print(X_train.size(), X_test.size(), Batches_X_train.size(), Batches_X_test.size())
```
Comment (JacksonMaxfield) [inline CVAE_testbed/datasets/quilt_aics_features.py]: Again, prefer `logging` over `print` but I don't know how you have your `bin` script setup

File: CVAE_testbed/datasets/quilt_aics_features.py [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+
+            Batches_X_train = torch.cat([Batches_X_train, X_train], 0)
+            Batches_C_train = torch.cat([Batches_C_train, C_train], 0)
+            Batches_conds_train = torch.cat([Batches_conds_train, conds_train], 0)
+            try:
+                Batches_X_test = torch.cat([Batches_X_test, X_test], 0)
+                Batches_C_test = torch.cat([Batches_C_test, C_test], 0)
+                Batches_conds_test = torch.cat([Batches_conds_test, conds_test], 0)
+            except:
+                pass
```
Comment (JacksonMaxfield) [inline CVAE_testbed/datasets/quilt_aics_features.py]: But... What is it catching? And the error is okay to just let through?

Comment (JacksonMaxfield) [reply CVAE_testbed/datasets/quilt_aics_features.py]: This is the second time I have seen this... Generally, you should always specify which exception it is catching. If you have multiple exceptions you are catching just place them in a tuple. This just good practice but also makes it easier to write tests later on.

File: CVAE_testbed/datasets/quilt_aics_features.py [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+        return len(self._batches_x)
+
+    def __getitem__(self, idx):
+        """
+        Returns a tuple. (X, C, sum(C[mid:end])). 
+        X is the input, 
+        C is the condition, 
+        sum(C[mid:end]) is the sum of the indicators in C. It tells us how many of the condition
+        columns have been masked
+        """
```
Comment (JacksonMaxfield) [inline CVAE_testbed/datasets/quilt_aics_features.py]: Docstrings 😍

File: CVAE_testbed/datasets/quilt_aics_features.py [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+    def get_train_data(self):
+        return self._batches_x_train, self._batches_c_train, self._batches_conds_train
+
+    def get_test_data(self):
+        return self._batches_x_test, self._batches_c_test, self._batches_conds_test
+    
+    def __len__(self):
+        return len(self._batches_x)
+
+    def __getitem__(self, idx):
```
Comment (JacksonMaxfield) [inline CVAE_testbed/datasets/quilt_aics_features.py]: Wait a second... you have two `__getitem__` functions? How does Python not complain about this...

File: CVAE_testbed/datasets/synthetic.py [hunk 12]
```diff
@@ -37,8 +37,9 @@ def __init__(self, num_batches, BATCH_SIZE, model_kwargs, shuffle=True, corr=Fal
             if self.shuffle is True:
                 while count == 0:
                     C_mask = torch.zeros(C.shape).bernoulli_(0.5)
-                    if len(set([i.item() for i in torch.sum(C_mask, dim = 1)])) == self.model_kwargs['x_dim'] + 1:
-                        count = 1 
+                    count = 1
```
Comment (JacksonMaxfield) [inline CVAE_testbed/datasets/synthetic.py]: So this will only run once?

File: CVAE_testbed/datasets/quilt_aics_features.py (hunk 3 above)
Comment (ritvikvasan) [reply CVAE_testbed/datasets/quilt_aics_features.py]: right!

File: CVAE_testbed/datasets/quilt_aics_features.py (hunk 4 above)
Comment (ritvikvasan) [reply CVAE_testbed/datasets/quilt_aics_features.py]: trying to read json files with keys not in ['imsize_orig', 'com', 'angle', 'flipdim', 'imsize_registered' ] and also trying to add CellID to this new dictionary. So i get a "features plus cellid" dataframe.

File: CVAE_testbed/datasets/quilt_aics_features.py (hunk 6 above)
Comment (ritvikvasan) [reply CVAE_testbed/datasets/quilt_aics_features.py]: trying different normalizations, need to clean this up.

File: CVAE_testbed/datasets/quilt_aics_features.py (hunk 8 above)
Comment (ritvikvasan) [reply CVAE_testbed/datasets/quilt_aics_features.py]: agree. will move to logging, print is habit

File: CVAE_testbed/datasets/quilt_aics_features.py (hunk 9 above)
Comment (ritvikvasan) [reply CVAE_testbed/datasets/quilt_aics_features.py]: ill fix this

File: CVAE_testbed/datasets/quilt_aics_features.py (hunk 11 above)
Comment (ritvikvasan) [reply CVAE_testbed/datasets/quilt_aics_features.py]: huh

File: CVAE_testbed/datasets/synthetic.py (hunk 12 above)
Comment (ritvikvasan) [reply CVAE_testbed/datasets/synthetic.py]: This generates multiple C_masks (each with a 50% probability of zeros) until the mask includes all possible permutations of 0's and 1's

File: CVAE_testbed/datasets/quilt_aics_features.py (hunk 4 above)
Comment (JacksonMaxfield) [reply CVAE_testbed/datasets/quilt_aics_features.py]: Got it. I personally think this would be a bit more understandable if you just include them all and then after simply:
```python
# Remove keys we don't care about
for key in ['imsize_orig', 'com', 'angle', 'flipdim', 'imsize_registered']:
   a.pop(key)
```

Comment (JacksonMaxfield) [reply CVAE_testbed/datasets/quilt_aics_features.py]: But otherwise I understand it so maybe just a comment if you don't like the above ^
